### PR TITLE
Change the main page to a list of weeks and add a separate page for entering shopping costs.

### PR DESCRIPTION
**Pull request by bighuggies**
Major change is switching to 1 graphql query for a page.

The whole graph object is refreshed whenever any action is taken and there is no paging on weeks. Could get slow, might need paging later but for now should be k

New main page (images are placeholders from placeimg):

localhost-4200- 1

New shopping page:

localhost-4200- 2

### CODE REVIEW

**Comment by bighuggies** [comment]
Main page when you need to pay:

localhost-4200- 3

**Comment by FraserKillip** [comment]
LGTM

**Review by ShadowDog007 — COMMENTED** [review]
What about showing who joined the week?

**Comment by bighuggies** [comment]
I didn't think seeing who joined was that useful,. I was planning on adding that to the shopping page so Lars can see who hasn't paid though.

**Comment by ShadowDog007** [comment]
Yeah, I think that's my main concern, an easy way to see who was in a given week so we can chase them up about it.

But if the group as a whole can see that then it could be easier?

I'm not really concerned about the backend performance at this point, our graphql api returns in ~50ms (1sec to validate new facebook tokens q_q)
I'm also in-progress of solving the N+1 problem.

Are there any extra fields which would make this easier/faster? Such as a user count?

I was thinking we could have a archiving strategy to clear out old data, e.g. filter weeks which have no pending debt and are a few weeks old?

File: src/app/weeks/weeks.component.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  templateUrl: './weeks.component.html',
+  styleUrls: ['./weeks.component.less']
+})
+export class WeeksComponent implements OnInit {
+
+  // we don't use thisweek but if we don't request it then new weeks will
+  // never be created in the db.
+  private weeksGql = gql`
+    query {
+      thisweek {
```
Comment (ShadowDog007) [inline src/app/weeks/weeks.component.ts]: You don't use thisweek ?

Also note that it won't match up with the frontend's 'current week'

Comment (bighuggies) [reply src/app/weeks/weeks.component.ts]: Is the reason in the comment not correct?

File: src/app/weeks/weeks.component.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    this.weeksQuery = this.apolloClient.watchQuery({ query: this.weeksGql });
+
+    this.weeksQuery.subscribe(({ data, loading }) => {
+      this.loading = loading;
+      this.me = data.me;
+      this.primaryShopper = data.primaryShopper;
+
+      const sortedWeeks = sortBy(data.weeks, 'weekId').reverse();
+
+      this.activeWeeks = sortedWeeks.filter(w => !this.isDueWeek(w.weekId));
```
Comment (ShadowDog007) [inline src/app/weeks/weeks.component.ts]: You should specifically fetch these weeks, listing all weeks only lists weeks which are in the database

File: src/app/shopping/shopping.component.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    this.loading = true;
+
+    this.currentWeekId = this.weekService.getCurrentWeekId();
+    this.weeksQuery = this.apolloClient.watchQuery({ query: this.shoppingGql });
+
+    this.weeksQuery.subscribe(({ data, loading }) => {
+      this.loading = loading;
+      this.me = data.me;
+      this.primaryShopper = data.primaryShopper;
+      this.weeks = sortBy(data.weeks, 'weekId').reverse();
```
Comment (ShadowDog007) [inline src/app/shopping/shopping.component.ts]: You should specifically request the current two weeks, or inject them in if they are not there.
Backend only lists weeks which are in the database.

File: src/app/weeks/weeks.component.ts (hunk 2 above)
Comment (ShadowDog007) [reply src/app/weeks/weeks.component.ts]: You could also inject defaults in for missing items that we want.

File: src/app/weeks/weeks.component.ts (hunk 1 above)
Comment (ShadowDog007) [reply src/app/weeks/weeks.component.ts]: No, weeks are deleted if they don't have a shopper || a cost.

We do the same for WeekUserLinks

**Comment by bighuggies** [comment]
I think this looks ok to show the members joined for a week (the avatars have hover tooltips):

image

If you added a field to a week which was the weekLink for the current user that would make things a little easier. At the moment I'm searching all the weekLinks for a week for one with a userId matching the current user.

**Comment by bighuggies** [comment]
Added some more stuff to Lars' page:

image

**Comment by ShadowDog007** [comment]
Would it be better to show amounts owed per person in a list instead of against weeks?

I want to move payments to just be 'I paid this amount at this time' instead of binding them to a week.

I can make a field for this to make it easier?

File: src/app/week-list/week-list.component.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<div *ngFor="let week of weeks" class="week">
+<md-card class="week">
+  <md-card-header>
+    <img md-card-avatar src="https://placeimg.com/40/40/animals" />
```
Comment (ShadowDog007) [inline src/app/week-list/week-list.component.html]: Wait what....??

**Comment by bighuggies** [comment]
image

Added amount owed per member below shopping table

File: src/app/shopping/shopping.component.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            avatarUrl
+          }
+        }
+        shopper {
+          firstName
+          lastName
+        }
+      }
+      users {
+			  userId
```
Comment (ShadowDog007) [inline src/app/shopping/shopping.component.ts]: tabs?

**Review by ShadowDog007 — APPROVED** [review]

**Pull request merged**